Create README.md

sachendras opened the pull request.

robshakir commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,382 @@
+
+   * Test Steps:
+       * Please run all the health checks suggested above prior to starting the test as a baseline.
+       * Configure DUT as suggested above.
```
```suggestion
       * Configure DUT as described above.
```

robshakir commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,382 @@
+           * Dont start the above mentioned encaped flows from ATE2:Port3<br>
+           
+   * Expectations:
+	    * ATE1:Port1 to ATE2:Port1 flows are expected to be 100% successful and routing over the IBGP peering between DUT:Port2 and ATE2:Port1
```
Can we define 100% successful? I assume 0% packet loss is your pass/fail criteria?

robshakir commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,382 @@
+
+
+**PF-1.6.2: Migrate routing of BE1 flows from ATE1:Port1 to ATE2:Port1 to ATE2:Port3 instead**
+   * Situation:
```
In this scenario -- I think it has become too complex defining the prefixes to selectively advertise in each test case. I'd suggest that we break this down in terms of the prefixes that we have in all scenarios, and then the deltas that are being created per test case.

The alternative is to define these prefixes in terms of sets (and the sessions with a name) and say, for example:

> ### Definitions:
> * `FLOW-SET-1`: ATE port 1 to ATE port 2
> * `BGP-SESSION-1`; ATE port 1 to ATE port 2, iBGP with AS100, policy setting local pref to 42 for all routes.
>
> ### Test Case
> Enable FLOW-SET-1 and BGP-SESSION-1.

I think also defining this in terms of the positive (i.e., enable flow 1) rather than the negative (i.e., disable flow 2) is clearer, because the reader can then just understand that they enable the list of flows that you specified. I can imagine writing this like:

```golang

flows := map[string]*someFlowThingy{
  "flow1": {
     src: "ATE-1",
     dst: "ATE-2",
     // some spec of the flow
  },
  "flow2": {
    src: "ATE-3",
    dst: "ATE-4",
  },  
}

func TestSomething(t *testing.T) {
   enableFlows(t, "flow1", "flow2", ...)
}
```

which is pretty easy to read.

sachendras replied:
Address this concern. Ptal at the defined tables. Hopefully that addresses the concern here.

robshakir commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,382 @@
+           * Ensure that the TTL on the outer heder is decremented and lower than that of the inner header.
+           * TOS byte in the outer header is the same as the TOS value in the inner header.
+           * Ensure the DUT is streaming accurate data on the number of packets/bytes being Encaped.
+	   * ATE2 to ATE1:Port1 Unencaped flows must be 100% successful. The flows are expected to route through the IBGP peering between ATE2:Port1 --> DUT:Port2 --> ATE
```
This is helpful as a clarification -- since it gives you a clear idea of what to do code w.r.t correctness criteria. 

If you don't agree w.r.t specifying the flows per subtest, then I suggest that we just define the flows in terms of src/dst IP and then define the correctness criteria like you have here.

sachendras replied:
I added flow-sets so resolving the comment

robshakir commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,382 @@
+
+**PF-1.6.3: Migrate routing of AF1-AF4 flows from ATE1:Port1 to ATE2:Port1 to ATE2:Port3 instead**
+   * Follow the same steps as above in PF-1.6.2 and gradually move one Traffic class at a time in the following order
+       * PF-1.6.3: Migrate routing of AF1 flows from ATE1:Port1 to ATE2:Port1 to ATE2:Port3 instead
```
Can we clarify the implementation here? It seems like we are doing this by advertising a particular prefix one at a time to on the session @ ATE2:port3, right?

We should clarify what steps in 1.6.2 we are covering here -- since migrating per AFI isn't defined it's not really clear what to do here (I think you're adding to the iBGP session advertisements, but it's not clearly specified).

sachendras replied:
lot of the comments may be stale at this point given the significant changes to the Readme so far.

robshakir commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,382 @@
+   * Test Steps:
+       * Please run all the health checks suggested above prior to starting the test as a baseline.
+       * Start all flows above:
+           * Dont start the above mentioned Unencaped flows from ATE2:Port1<br>
```
Again, let's define the positive case by having some way to group the flows.

sachendras replied:
Marking this resolved since the flow-sets are defined now.

robshakir commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,382 @@
+       * Flows from ATE1 to ATE2 will be GUE encaped to tunnel destination IPv4Prefix12/28 and routed out to ATE2:Port3. The flows must be successful. Ensure that the TTL on the outer heder is decremented and lower than that of the inner header. TOS byte value in the outer header is the same as the TOS value in the inner header.
+       * Ensure the DUT is streaming accurate data on the number of packets/bytes being Encaped.
+       * ATE2:Port3 to ATE1:Port1 Encaped flows with tunnel destination as IPv4Prefix11/28 must be 100% successful. The flows are expected to route through the IBGP peering between ATE2:Port1 --> DUT:Port2 --> ATE
+       * Ensure the DUT is streaming accurate data on the number of packets/bytes being Decaped.
```
Can we define the counter that we expect to use for this?

robshakir commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,382 @@
+   * Test Steps:
+       * Please run all the health checks suggested above prior to starting the test as a baseline.
+       * Start all flows above:
+           * Dont start the above mentioned Unencaped flows from ATE2:Port1
```
Same comment as above re: negative vs. positive definitions. Will not comment on the remainder, but I'd suggest we make these changes throughout.

robshakir commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,382 @@
+           * Dont start the above mentioned Unencaped flows from ATE2:Port1
+       * Configure the EBGP on ATE2:Port3 to stop advertising "IPv4Prefix12/28" to DUT:Port4. When this happens, following is expected.
+           * The Static route that points the VPNH to the tunnel destination (NHG or tunnel interface) must become invalid. Therefore, all the routes that are programmed by ATE2:Port2 with a VPNH of "IPv4Prefix17" in its IBGP session with DUT:Port3 must become invalid.
+           * Because of the above, all IBGP routes recevied by the DUT:Port2 from ATE2:Port1 with the default local-preference MUST become valid and hence there must not be any traffic loss.<br><br><br>
```
```suggestion
           * Because of the above, all IBGP routes received by the DUT:Port2 from ATE2:Port1 with the default local-preference MUST become valid and hence there must not be any traffic loss.<br><br><br>
```

sachendras replied:
This comment is no longer valid, so resolving. The test is rewritten since the last review

robshakir commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,382 @@
+			
+
+   * ATE2:Port2 <-IBGP-> DUT:Port3 over IPv4-unicast and IPv6-Unicast
+       * ATE2:Port2 -> DUT:Port3. Following prefixes for different services advertised with local-preference 200 and a Virtual Protocol Next-Hop (VPNH) of IPv4Prefix17 for IPv4 destination prefixes and a VPNH of IPv6Prefix17 for IPv6 destination prefixes. [Please note: Following advertisements are made gradually as suggested in the subtests below]:
```
For `IPv4Prefix17` here -- is this a /32, or is this actually a prefix?

robshakir commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,382 @@
+       * Start all flows above:
+           * Dont start the above mentioned Unencaped flows from ATE2:Port1
+       * Configure the EBGP on ATE2:Port3 to stop advertising "IPv4Prefix12/28" to DUT:Port4. When this happens, following is expected.
+           * The Static route that points the VPNH to the tunnel destination (NHG or tunnel interface) must become invalid. Therefore, all the routes that are programmed by ATE2:Port2 with a VPNH of "IPv4Prefix17" in its IBGP session with DUT:Port3 must become invalid.
```
It'd be useful in these test cases to define the configuration -- the static route here is specifically:

> static
>   dst: `IPv4Prefix17` 
>   nhop: `IPv4Prefix12`   # what address is actually the NH? this is a /28

so `IPv4Prefix12/28` is withdrawn, and `IPv4Prefix17` (which I assume is a /32) which is the VPNH becomes invalid, and hence at that point prefixes for whom the next-hop is set to `IPv4Prefix17` become invalid.

Given this scenario, you use the alternate routing.

It seems worth clarifying this because it's very easy to get lost in what the scenario here is IMHO.

sachendras replied:
I have addressed both the requirements. Ptal

sachendras replied:
I have revamped the test significantly to improve readability. This comment may be stale at this point so resolving. Please feel free to re-open.

robshakir commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
````diff
@@ -0,0 +1,382 @@
+```yaml
+paths:
+  # Policy for route leaking between VRFs
+  /network-instances/network-instance/inter-instance-policies/apply-policy
````
I think this isn't needed, there are no VRFs mentioned?

sachendras replied:
Yes, updated, ptal

robshakir commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,382 @@
+  # Following applied in Non-default VRF
+  /network-instances/network-instance/inter-instance-policies/apply-policy/config/import-policy
+
+  # Config for VRF selection
```
No VRF selection policies mentioned in this README AFAICS?

sachendras replied:
Updated, ptal

robshakir commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,382 @@
+  /routing-policy/defined-sets/prefix-sets/prefix-set/prefixes/prefix/config/ip-prefix:
+  /routing-policy/defined-sets/prefix-sets/prefix-set/prefixes/prefix/config/masklength-range:
+
+  # Apply VRF selection policy as per the Test procedure above
```
Assume this whole section around VRF selection is outdated.

robshakir commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,382 @@
+
+## Required DUT platform
+
+* Specify the minimum DUT-type:
```
Again, I think this is not the minimum -- must be MFF since MFF > FFF if you need MFF, but I'm not sure why this can't be executed against FFF.

sachendras replied:
I think that was an error, will correct it

sachendras replied:
Corrected, ptal

robshakir reviewed (changes requested):
I would suggest that this test case needs a diagram that explains where the routing is attached -- this can be added as an image that is then embedded in the README. 

I think it needs to show:

 * each group of prefixes that are advertised
 * where the BGP sessions are between -- interface vs. loopback addresses
 * where the IS-IS adjacencies are

It's quite hard to follow what is happening in each stage without a clearer topological idea of the scenario. 

I think for these combinatorial tests we have to be much clearer about the way the test setup should look.

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,382 @@
+# PF-1.6 Static GUE Encap and BGP path selection
+
+## Summary
```
ack, thanks!

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,382 @@
+# PF-1.6 Static GUE Encap and BGP path selection
+
```
Done, thanks

dplore commented:
@sachendras please respond to and resolve open comments. Thanks!

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,382 @@
+	
+   * ATE2:Port1 <-IS-IS(L2)-> DUT:Port2
+       * ATE2:Port1 --> DUT:Port2
+           * IPv4Prefix13, IPv6Prefix13 <<--- ATE2:Port1 IBGP address
```
ack on that will address it

sachendras replied:
I have addressed this requirement and added tables too for all the IPs used in the test

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,382 @@
+           * IPv4Prefix9/24 IPv6Prefix9/64 <--- Advertised in PF-1.6.5 to PF-1.6.9
+           * IPv4Prefix10/24 IPv6Prefix10/64 <--- Advertised in PF-1.6.6 to PF-1.6.9
+       * DUT:Port3 --> ATE2:Port2
+           * All Intternet Prefixes learnt from ATE1:Port1
```
ack

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,382 @@
+### DUT configuration:
+   * EBGP peering between DUT:Port1 and ATE1:Port1
+   * IS-IS adjancency between DUT:Port2 and ATE:Port1
+       * Loopback interface is passive.
```
ack

sachendras replied:
Updated this in the test

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,382 @@
+   * IBGP peering between DUT:Por2 and ATE2:Port1
+   * IBGP peering between DUT:Port3 and ATE2:Port2
+   * EBGP peering between DUT:Port4 and ATE2:Port3
+   * Import and export route policies to match the advertisements for each BGP peering.
```
I will add some clarification

sachendras replied:
The  ambiguity doesnt exist anymore as the test is rewritten with added details and table references. The new "BGP Peering Details" table explicitly defines which prefixes are advertised from the ATE side in each session

masood-shah commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,584 @@
+
+    *   For a GUE decapsulation node, configure the following:
+        *   UDP port 6080 must be used for decapsulating IPv4 payload, and UDP
+            port 6615 for IPv6 payload
```
@sachendras  I think (today) the decision was/is to use a single, designated UDP destination port for gIRO's IP/UDP encap/decap, accommodating both inner IPv4 and IPv6 payloads. This same UDP port indicates the inner payload is IP; and the first nibble of that inner payload will then be used to differentiate between IPv4 and IPv6. The decapsulated packet will subsequently be forwarded based on its inner IP. Should we proceed with updating the documentation to reflect this?

sachendras replied:
I can correct that. This says the old intent since I wasnt clear on support

masood-shah replied:
Since we've decided this is the way forward, let's change it and start working with vendors to get it codified

masood-shah commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,584 @@
+
+    *   Configure static GUE encapsulation as follows:
+        *   Define UDP ports to be used for IPv4oUDP and IPv6oUDP as 6080 and
+            6615 respectively.
```
Ditto as above

masood-shah commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,584 @@
+1.  Test the implementation of Static GUE encap where the Tunnel endpoint is
+    resolved over EBGP, while the Payload's destination is learned over IBGP.
+
+2.  Verify that before GUE encapsulation, the LPM lookup on the payload
```
We can omit 'LPM' here and simplify the text to focus on the intended result...For example: "Confirm that prior to encapsulation, the DUT selects the correct path to the payload destination by prioritizing the IBGP route with the highest Local Preference, and successfully utilizes backup IBGP routes if the primary path fails."

masood-shah commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,584 @@
+    the implementation marks the TZ bits (LSB bit \#1 and 2 of the DSCP header)
+    as either TZ=11 or TZ=10.
+
+4.  Ensure that the implementation allows configuration of the TTL value for the
```
Refine it slightly to enhance clarity...such as..."Confirm that the TTL value for the outer IP header created during GUE encapsulation can be explicitly configured."

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,382 @@
+       * IPv6Prefix10/64 to IPv6Prefix5/64 [DSCP:AF4]
+<br><br><br>
+
+### Following Health checks to be run Pre and Post every sub test
```
This is a placeholder for the list of different pre and post tests. In each of the tests below, I have called this placeholder pre and post every test to ensure the test didnt result in any of the errors/drops. Let me know if my understanding is incorrect here.

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,382 @@
+
+### Advertisements:
+
+   * ATE1 <-EBGP-> DUT:Port1 over IPv4-unicast and IPv6-Unicast
```
ok, I will address these

masood-shah commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,616 @@
+1.  Test the implementation of Static GUE encap where the Tunnel endpoint is
+    resolved over EBGP, while the Payload's destination is learned over IBGP.
+
+2.  Verify that before GUE encapsulation, the LPM lookup on the payload
```
@sachendras suggesting a slight rewrite, removing LPM, something like this..."Confirm that before GUE encapsulation, the device correctly selects the path for the payload destination from multiple available IBGP routes. Path selection needs to follow the BGP best path algorithm rules, like giving preference to routes with a higher BGP Local Preference. Furthermore, it must successfully switch over to backup/alternative IBGP paths when the preferred path fails"

sachendras replied:
Done

masood-shah replied:
ty!

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,627 @@
+| `$DUT_TE10.v4/32`                                                                                          | IPoUDP tunnel destination address on the DUT. This IP MUST receive traffic with TE bits marked as 10                                                                                                                                                                                             ...
+| `$ATE2_INTERNAL_TE11.v4/32`                                                                                | IPoUDP tunnel destination address on the ATE2. This IP MUST receive traffic with TE bits marked as 11                                                                                                                                                                                            ...
+| `$ATE2_INTERNAL_TE10.v4/32`                                                                                | IPoUDP tunnel destination address on the DUT. This IP MUST receive traffic with with TE bits marked as 10                                                                                                                                                                                        ...
+| <br>`$ATE2_INTERNAL6.v4/24`<br>`$ATE2_INTERNAL6.v6/64`<br>`$ATE2_INTERNAL7.v4/24`<br>`$ATE2_INTERNAL7.v6/24`<br>`$ATE2_INTERNAL8.v4/24`<br>`$ATE2_INTERNAL8.v6/64`<br>`$ATE2_INTERNAL9.v4/24`<br>`$ATE2_INTERNAL9.v6/24`<br>`$ATE2_INTERNAL10.v4/24`<br>`$ATE2_INTERNAL10.v6/64` | <br>- Internal Public prefixes<br>- Advertised to the DUT over the IBGP peering between `ATE2_PORT1.IBGP.v[46]`<>`$DUT_lo0 ...
```
That was stale info, corrected now. Thanks for spotting

dplore commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
IMO, this test could be broken up into several, smaller, separate tests:

1. static route to BGP NH with IP in GUE encapsulation 
1. static route to BGP NH with IP in GUE decapsulation
1. qos for static route to BGP NH with IP in GUE encap. (depends on 1.)
1. qos for static route to BGP NH with IP in GUE decap (depends on 2.)
1. scale qos for static route to BGP NH with IP in GUE encap (depends on 3.)
2. 1. scale qos for static route to BGP NH with IP in GUE decap (depends on 4.)

We used this pattern of building up a base encap test and then layering on qos and scale  for the policy-forwarding tests for MPLS in GUE encap/decap with success

sachendras replied:
Is the ask to have separate FNTs for each?
If so, there are FNTs already in review that are covering smaller portions that you mentioned. However here I want multidimensional test as the title suggests.

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,635 @@
+4.  Confirm that the TTL value for the outer IP header created during GUE
+    encapsulation can be explicitly configured.
+
+5.  Confirm that the DUT performs GUEv1 Decap of traffic received in the reverse
+    direction, and that during decapsulation, the outer DSCP and TTL bits are
+    not copied to the inner header.
```
Why is this being called out now? You did the review of the PR last year too https://github.com/openconfig/featureprofiles/pull/3486#issuecomment-2557882672

Also, the requirement here is to have Multidimensional test. So I prefer to have the test multidimensional. Single dimension tests are expected to be covered in other Pulls.

dplore commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,635 @@
+
+In addition to the adjacencies and peering configurations described in the
+tables above, the DUT requires the following configurations:
+
```
here's another example where one test refers to a second test to "import"  or re-use the configuration.  

https://github.com/openconfig/featureprofiles/blob/fccde73b7301161320faf1afd431d26b8976b531/feature/policy_forwarding/otg_tests/mpls_gre_ipv4_encap_scale_test/README.md?plain=1#L29

sachendras replied:
The canonical JSON is added here so I dont see a need for reference. Resolving this comment.

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,635 @@
+        baseline. Verify that there are no drops, core dumps, or other issues.
+
+
+**RT-3.52.2: BE1 Traffic Migrated from being routed over the DUT_Port2 -->
```
Sure, but this one I wanted a multidimensional test covering functionality of a certain set of features in tandem.

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,635 @@
+
+        
+
+**RT-3.52.7: DUT as a GUE Decap Node**
```
I have several smaller tests for each feature in isolation. As mentioned earlier, I want to keep this one as a multidimensional test.

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,635 @@
+| ASN2                                 |                      | `$DUT_Port2`                                                                                                                                                                                                                                                                                                                                    ...
+
+**Different Flows used throughout the test:**
+You are correct, the previous markdown table was inaccurate. Here's the rebuilt markdown version:
```
yep, corrected

masood-shah reviewed (commented):
LGTM

dplore commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,635 @@
+
+    *   Flow-Set #1 and Flow-Set #2 are expected to achieve 100% success and be
+        routed over the connection between `DUT_Port2` <> `ATE2_Port1`. Since no
+        encapsulation is performed on the DUT in this scenario, the TZ bits are
```
You'll need to define tz bits as this is not standard

dplore replied:
Oh I see an explanation was added

dplore commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,635 @@
+    *   Execute all health checks mentioned above prior to running the tests.
+    *   Configure the DUT as described previously for IS-IS adjacency, IBGP and
+        EBGP peering and the corresponding routing policies, static routes for
+        GUE encapsulation, and policies to selectively apply Traffic Class (TZ)
```
Thanks!

sachendras commented:
ack, added a diagram as you suggested and also embedded it in to the readme. While I have all the protocols, adjacencies and the IP addresses used for the peering detailed in the diagram, I have also added multiple tables to the test explaining what addresses are advertised by the different peers, Ptal

sachendras commented:
/gcburn

sachendras commented:
/gcbrun

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,706 @@
++-------------------------------------------+            |          |
+                                                         |          |
++-----------------------------------------------------+  |          |
+|                      ATE2 (ASN 100 & 200)           |  |          |
+|-----------------------------------------------------+  |          |
+| Port1: (to DUT)                                     <--+          |
+|  - $ATE2PORT1_DUTPORT2_1.v4/31                      |             |
+|  - $ATE2PORT1_DUTPORT2_1.v6/127                     |             |
+|-----------------------------------------------------+             |
+| Port3: (to DUT)                                     <-------------+
+|  - `$ATE2PORT3_DUTPORT4_1.v4/31`                    |
+|  - `$ATE2PORT3_DUTPORT4_1.v6/127`                   |
+|-----------------------------------------------------+
```
I added a statement that ATE1 and ATE2 can be the same device.

dplore commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,709 @@
+*   **Test Steps:**
+
+    *   Execute the health checks described previously
+    *   Configure the DUT according to the prior instructions
```
Has the configuration changed?   The DUT config defined in the setup was pushed in 3.52.1 already.  Do you want to change the DUT config in some way and push it again?

sachendras replied:
no, I updated.. it is the same as the baseline config,

dplore commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,709 @@
+**RT-3.52.3: AF1-AF4 Traffic Migrated to DUT --> ATE2_Port3**
+
+Follow the same steps as in RT-3.52.2 and gradually move one Traffic class at a
+time in the following order. Note changes in RT-3.52.5 and RT-3.52.6
```
specify the config needed to move traffic

dplore commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,709 @@
+        destinations `$ATE2_INTERNAL_TE11.v4/32` and `$ATE2_INTERNAL_TE10.v4/32`
+        and routed out `$DUT_Port4<>$ATE2_Port3`.
+
+    *   The DUT should provide accurate streaming data for the number of
```
specify OC paths needed for this (here and in the OC Paths at the yaml stanza at end of this README)

sachendras replied:
Done

dplore commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,709 @@
+
+        Traffic should reach the destination successfully.
+
+    *   The DUT should accurately stream data on decapsulated packets/bytes.
```
specify OC paths needed for this (here and in the OC Paths at the yaml stanza at end of this README)

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,709 @@
+
+*   **Situation:**
+
+    *   The baseline test (RT-3.52.1) is running, with flows active in
```
Addressed, asked to restart the flows.

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,709 @@
+    *   The outer header DSCP bits should be the same as the inner header DSCP
+        bits when received at `ATE2_Port1` (before decapsulation).
+    *   The DUT should accurately stream data regarding the number of
+        packets/bytes encapsulated
```
Added the paths and also clarified which parts to be validated in ATE v/s which parts in OC

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,743 @@
+|  - $ATE1_PORT1_INTERNET[1-5].v6/64        |
+|                                           |
+| ATE1_Port1 <-> DUT_Port1:                 |
+|  - $ATE1PORT1_DUTPORT1_1.v4/31            |
```
Well, intention is to have 
$ATE1PORT1_DUTPORT1/31 as the prefix in which

$ATE1PORT1_DUTPORT1_1/32 is the ATE IP and $ATE1PORT1_DUTPORT1_2/32 is the DUT IP.

Which is why in the diagram, the port config on the ATE and the DUT sides carry their respective IPs from the same subnet.

ram-mac replied:
ACK

sachendras replied:
They aren't, but 2 different IP address of the same /31 prefix. _1 represents ATE side of the IP whereas _2 is DUT side of the IP. Following from the All_IPaddresses table above.

<br>`$ATE1PORT1_DUTPORT1_1.v4/31`<br>`$ATE1PORT1_DUTPORT1_1.v6/127`                                                                                                                                                                                                                                                                        | Configured on the ATE1PORT1 port of ATE1_Port1<>DUT_Port1 point-to-point connection and used for IS-IS adjacency.
<br>`$ATE1PORT1_DUTPORT1_2.v4/31`<br>`$ATE1PORT1_DUTPORT1_2.v6/127`                                                                                                                                                                                                                                                                        | Configured on the DUTPORT1 port of ATE1_Port1<>DUT_Port1 point-to-point connection and used for IS-IS adjacency.

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,743 @@
+| Flow-Set#4 from ATE2_Port3 --> ATE1_Port1 are GUE encaped with Tunnel destination as $DUT_TE10.v4/32 | `$ATE2_INTERNAL9.v6/64 --> $ATE1_PORT1_INTERNET4.v6/64`  | AF3  | `$DUT_TE10.v4/30`           |
+| Flow-Set#4 from ATE2_Port3 --> ATE1_Port1 are GUE encaped with Tunnel destination as $DUT_TE10.v4/32 | `$ATE2_INTERNAL10.v4/24 --> $ATE1_PORT1_INTERNET5.v4/24` | AF4  | `$DUT_TE10.v4/30`           |
+| Flow-Set#4 from ATE2_Port3 --> ATE1_Port1 are GUE encaped with Tunnel destination as $DUT_TE10.v4/32 | `$ATE2_INTERNAL10.v6/64 --> $ATE1_PORT1_INTERNET5.v6/64` | AF4  | `$DUT_TE10.v4/30`           |
+| Flow-Set#5 from ATE2:Port1 to ATE1:Port1 are sent Unencaped                                          | `$ATE2_INTERNAL6.v4/24 --> $ATE1_PORT1_INTERNET1.v4/24`  | BE1  | N/A                         |
```
ack, updated all the internet prefixes with user prefixes. Hopefully that helps?

So, the following,
- $ATE1_PORT1_INTERNET[1-5].v4/24
- $ATE1_PORT1_INTERNET[1-5].v6/64 

will change to:
- $ATE1_PORT1_user[1-5].v4/24
- $ATE1_PORT1_user[1-5].v6/64

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,743 @@
+
+| Src_destination of flows | From_IP --> To_IP | DSCP | Tunnel endpoint used |
+| :------------------------------------------------------------------------------------------------- | :------------------------------------------------------ | :--- | :-------------------------- |
+| Flow-Set#1 from ATE1_Port1 --> ATE2_[Either Port1 or Port3 depending on the FIB entries of the DUT]  | `$ATE1_PORT1_INTERNET1.v4/24 --> $ATE2_INTERNAL6.v4/24`  | BE1  | `ATE2_INTERNAL_TE11.v4/32`  |
```
Not sure I follow, below is from a snippet of the `All IP addresses for the test:`  table. Also, I have the same captured under the "Topology" section plus the IBGP table for advertisements. Are you referring to certain section of the doc?

<br>`$ATE2_INTERNAL6.v4/24`<br>`$ATE2_INTERNAL6.v6/64`<br>`$ATE2_INTERNAL7.v4/24`<br>`$ATE2_INTERNAL7.v6/64`<br>`$ATE2_INTERNAL8.v4/24`<br>`$ATE2_INTERNAL8.v6/64`<br>`$ATE2_INTERNAL9.v4/24`<br>`$ATE2_INTERNAL9.v6/64`<br>`$ATE2_INTERNAL10.v4/24`<br>`$ATE2_INTERNAL10.v6/64`                                                           | <br>- Internal Public prefixes<br>- Advertised to the DUT over the IBGP peering between `ATE2_PORT1.IBGP.v[46]`<>`$DUT_lo0.v[46]` and `$ATE2_C.IBGP.v6`<>`$DUT_lo0.v6`<br>- Advertised further to `$ATE1_IBGP.v[46]` over the IBGP peering between `$ATE1_IBGP.v[46]`<>`DUT_lo0.v[46]` over their respective AFI peering

ram-mac replied:
Yes, i thought the Internal prefixes are reference from other documents internally. But now it is clear that it is referred to as internally to the README itself.

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,743 @@
+|  - $DUT_lo0.v6                            |
+|-------------------------------------------|
+| Port2: (to ATE2_Port1)                    <------------+ (IS-IS/IBGP)
+|  - $ATE2PORT1_DUTPORT2_2.v4/31            |            |
```
It is not OR, I corrected that to "and". Hopefully it helps now? Basically IS-IS provides reachability for IBGP peering.

ram-mac replied:
Thanks; This helps

ram-mac reviewed (changes requested):
@sachendras - There are some issues with OC paths in the README which is why it is failing the ci/cd checks. Can you fix them?

ram-mac reviewed (approved):
LGTM

robshakir reviewed (approved):
Approving based on @rohit-rp and @ram—mac’s reviews.

dplore commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
````diff
@@ -0,0 +1,745 @@
+    destination `$ATE2_INTERNAL_TE10.v4/32`.
+*   BE1-AF4 are now migrated
+*   Complete validation same as RT-3.52.2 above
+```
````
Need to remove this to get the README to validate

```suggestion
```

dplore commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
````diff
@@ -0,0 +1,745 @@
+    *   Use gnmi to push the configuration to the DUT
+    *   Ensure no prefixes are exchanged over the IBGP peering between
+        `$ATE2_C.IBGP.v6` and `$DUT_lo0.v6`. Validate this using OC
+        ```
````
The use of the triple backtick has to be on the 1st column, otherwise you get unpredictable results.  Notice how much of the remainder (but not all) is treated as static text.  

Convert all the triple back ticks to start and end on the 1st column and check the mdformat results to make sure this looks as intended.   

Right now major sections of the doc are getting excluded as static /codeblock text which is causing validation to fail for ocpaths and canonical_oc

sachendras replied:
Done

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
````diff
@@ -0,0 +1,745 @@
+        /network-instances/network-instance/afts/ipv6-unicast/ipv4-entry/state/prefix
+        ```
+
+## Canonical OpenConfig for GUEv1 Encapsulation configuration
````
Done

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,743 @@
+        should be GUE-encapsulated with tunnel destination
+        `$ATE2_INTERNAL_TE11.v4` and routed over the EBGP peering between
+        `$ATE2_Port3` and `$DUT_Port4`, and these flows must be 100% successful
+        (zero loss).
```
done

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
````diff
@@ -0,0 +1,743 @@
+```
+*   Migrate routing of AF1 flows from `DUT_Port2` --> `ATE2_Port1` to
+`DUT_Port4` --> `ATE2_Port3`.
+*   BE1 and AF1 are now migrated
````
This is already provided in context when you read through the test in its entirety on how to do the migration.

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md:
```diff
@@ -0,0 +1,743 @@
+
+### RT-3.52.8: Negative Scenario - EBGP Route for remote tunnel endpoints Removed
+
+Inflight
```
it is the continuation of the heading. Looks like got cut-off due to the 80 character limit. I corrected. Thanks!

sachendras commented:
/gcbrun

sachendras commented:
/gcburn

sachendras commented:
/gcbrun
